### PR TITLE
Use an app factory pattern for Response Operations UI

### PR DESCRIPTION
**Pull request by jcox-dev**
# Motivation and Context

Makes it easier to spin up an instance (or instances) of the Flask application on demand. Useful for testing and less dependency on setting everything up at import-time.

# What has changed

Changed to use a reproducible app factory pattern (better late than never!).

# How to test?

Unit tests 🍰 

# Links

- [TBL#138](https://trello.com/c/zdMOd6hC)
- http://flask.pocoo.org/docs/1.0/patterns/appfactories/

### CODE REVIEW

**Comment by jcox-dev** [comment]
[Python services to use app factory pattern (5)](https://trello.com/c/zdMOd6hC/138-python-services-to-use-app-factory-pattern-5)

**Comment by jcox-dev** [comment]
Looks bigger than it is ;)

**Comment by benjefferies** [comment]
@jcox-dev have you pushed to CF to check that the procfile still works?

File: run.py [hunk 1]
```diff
@@ -3,16 +3,15 @@
...
-
-from response_operations_ui import app  # NOQA # pylint: disable=wrong-import-position
+from response_operations_ui import create_app
 
 
 logger = wrap_logger(logging.getLogger(__name__))
 
 
 if __name__ == '__main__':
+    if not os.getenv('APP_SETTINGS'):
```
Comment (benjefferies) [inline run.py]: This should be ok to default to dev right because CF will call create_app() in CF?

Comment (jcox-dev) [reply run.py]: Precisely 👍 CF won't be using this file. Only docker and local dev.

File: response_operations_ui/error_handlers.py [hunk 2]
```diff
@@ -3,33 +3,36 @@
...
 def server_error_500(error):  # pylint: disable=unused-argument
     return redirect(url_for('error_bp.server_error_page'))
+
+
+def setup_error_handlers(app):
+    app.register_error_handler(ApiError, api_error)
+    app.register_error_handler(401, handle_authentication_error)
+    app.register_error_handler(Exception, server_error)
+    app.register_error_handler(500, server_error_500)
+    return app
```
[the remark this answers is not in the document]
Comment (jcox-dev) [reply response_operations_ui/error_handlers.py]: Related docs:
* [Custom Error Pages](http://flask.pocoo.org/docs/1.0/patterns/errorpages/#errorpages)
* [Error Handlers with Blueprints](http://flask.pocoo.org/docs/1.0/blueprints/?highlight=error%20handling#error-handlers)

[the remark this answers is not in the document]
Comment (jcox-dev) [reply response_operations_ui/error_handlers.py]: We could do something similar to eQ: https://github.com/ONSdigital/eq-survey-runner/blob/master/app/views/errors.py

[the remark this answers is not in the document]
Comment (ajmaddaford) [reply response_operations_ui/error_handlers.py]: I don't think I'd describe the annotation as magic, if we went with that logic we'd have no annotations anywhere. I'd say it removes a level of indirection when linking an exception to it's handler 🤷‍♂️

[the remark this answers is not in the document]
Comment (jcox-dev) [reply response_operations_ui/error_handlers.py]: Agreed. I think it'd be preferable to immediately see directly which handler is for which error code.

[the remark this answers is not in the document]
Comment (jcox-dev) [reply response_operations_ui/error_handlers.py]: Ah! I think this solves both cases: http://flask.pocoo.org/docs/1.0/api/#flask.Blueprint.app_errorhandler

[the remark this answers is not in the document]
Comment (jcox-dev) [reply response_operations_ui/error_handlers.py]: @ajmaddaford updated in [new commit](https://github.com/ONSdigital/response-operations-ui/pull/208/commits/b379336d68a2293fe16930ca385d0271ac0101ef).

**Pull request merged**